source/arch/encoding_simd_sse41.c

```diff
@@ -0,0 +1,396 @@
+     * We need to pack them together (and shift down) with a shuffle operation.
+     * Unfortunately the shuffle operation operates independently within each 128-bit lane,
+     * so we'll need to do this in two steps: First we compact dwords within each lane, then
+     * we do a dword shuffle to compact the two lanes together.
+
+     * 15 14 13 12 11 10 09 08 07 06 05 04 03 02 01 00 <- byte index (little endian)
+     * -- 09 0a 0b -- 06 07 08 -- 03 04 05 -- 00 01 02 <- data index
+     *
+     * We also reverse the order of 3-byte fragments within each lane; we've constructed
+     * those fragments in little endian but the order of fragments within the overall
+     * vector is in memory order (big endian)
+     */
+    const struct aligned256 shufvec_buf = {{
+        /* clang-format off */
+        /* MSB */
+        0xFF, 0xFF, 0xFF, 0xFF, /* Zero out the top 4 bytes of the lane */
+        2,  1,  0,
+        6,  5,  4,
+        10,  9,  8,
+        14, 13, 12,
+
+        0xFF, 0xFF, 0xFF, 0xFF, /* Zero out the top 4 bytes of the lane */
+        2,  1,  0,
+        6,  5,  4,
+        10,  9,  8,
+        14, 13, 12
+        /* LSB */
+        /* clang-format on */
+    }};
+    __m256i shufvec = _mm256_load_si256((__m256i const *)&shufvec_buf);
+
+    dwords = _mm256_shuffle_epi8(dwords, shufvec);
+    /*
+     * Now shuffle the 32-bit words:
+     * A B C 0 D E F 0 -> 0 0 A B C D E F
+     */
+    __m256i shuf32 = _mm256_set_epi32(0, 0, 7, 6, 5, 3, 2, 1);
+
+    dwords = _mm256_permutevar8x32_epi32(dwords, shuf32);
+
+    return dwords;
+}
+
+static inline bool decode(const unsigned char *in, unsigned char *out) {
+    __m256i vec = _mm256_loadu_si256((__m256i const *)in);
+    if (!decode_vec(&vec)) {
+        return false;
+    }
+    vec = pack_vec(vec);
+
+    /*
+     * We'll do overlapping writes to get both the low 128 bits and the high 64-bits written.
+     * Input (memory order): 0 1 2 3 4 5 - - (dwords)
+     * Input (little endian) - - 5 4 3 2 1 0
+     * Output in memory:
+     * [0 1 2 3] [4 5]
+     */
+    __m128i lo = _mm256_extracti128_si256(vec, 0);
+    uint64_t hi = _mm256_extract_epi64(vec, 2);
+
+    _mm_storeu_si128((__m128i *)out, lo);
+    memcpy(out + 16, &hi, sizeof(hi));
+
+    return true;
+}
+
+size_t aws_common_private_base64_decode_sse41(const unsigned char *in, unsigned char *out, size_t len) {
+    if (len % 4)
+        return -1;
+
+    size_t outlen = 0;
+    while (len > 32) {
+        if (!decode(in, out))
+            return -1;
+        len -= 32;
+        in += 32;
+        out += 24;
+        outlen += 24;
+    }
+
+    if (len > 0) {
+        unsigned char tmp_in[32];
+        unsigned char tmp_out[24];
+
+        memset(tmp_out, 0xEE, sizeof(tmp_out));
+
+        /* We need to ensure the vector contains valid b64 characters */
+        memset(tmp_in, 'A', sizeof(tmp_in));
+        memcpy(tmp_in, in, len);
+
+        size_t final_out = (3 * len) / 4;
```

Comment:
could go for a >> 2 here.

Reply:
Also will be handled by optimizing compilers.

Reply:
Similarly, is optimized away at -O1 ([godbolt](https://godbolt.org/z/_-fFos))